### DISCUSSION TITLE
C# 9: why is the compiler unable to infer the type when using target-typed new with chained methods?

### DISCUSSION THREAD

**Discussion by IanKemp** [dotnet/csharplang#5211] [Q&A] [answered]
Consider the following extension method:

```csharp
public static class ExtensionMethod
{
        public static THttpContext WithAuthorizationHeader<THttpContext>(this THttpContext httpContext, string value)
            where THttpContext : HttpContext
        {
            httpContext.Request.Headers.Add(HeaderNames.Authorization, value);

            return httpContext;
        }
}
```

The following code happily compiles, as expected:

```csharp
var context = new DefaultHttpContext().WithAuthorizationHeader($"Bearer {token}");
```

But this code **does not**, even though I would expect it to:

```csharp
DefaultHttpContext context = new().WithAuthorizationHeader($"Bearer {token}"); // CS8754 "There is no target type for 'new()'"
```

It appears the compiler is attempting to parse the entirety of the expression and infer the type from that, whereas to my mind it should be parsing and resolving `DefaultHttpContext context = new()` first, then evaluating the chained method call.

I have checked https://docs.microsoft.com/dotnet/csharp/language-reference/proposals/csharp-9.0/target-typed-new but can't see anything that would imply this shouldn't work.

**Comment by svick** [comment, marked as answer]
> to my mind it should be parsing and resolving `DefaultHttpContext context = new()` first, then evaluating the chained method call

Why would it do that? The compiler hasn't performed overload resolution yet, so it doesn't know that the type of the `new()` subexpression is going to be the same as the type of the whole expression. And without that, there is no connection between `new()` and `DefaultHttpContext`.

**Reply by jnm2** [reply]
I do find myself stopping and noticing this now and then and wondering made me expect it to work, since if it did work it would seem to behave unreliably.

**Reply by HaloFour** [reply]
It's interesting conceptually but it seems like it would be awfully complex.  It would effectively have to resolve all of the members in reverse, based on their return type.  That would potentially include extension methods or generic methods where normal inference would could this reverse resolution to break down.

**Reply by JimBobSquarePants** [reply]
Complex or not I strongly believe that the lack of support for this pattern delivers a very poor user experience for developers new to the language.

**Reply by TahirAhmadov** [reply]
This would be impossible. We don't know what `new()` is, so we don't know where `WithAuthorizationHeader` comes from. To make this work, the compiler would have to look at all members of all types in scope, including extensions, and find those with the expected return type, and still it would be error prone, because multiple members could exist that fit that criteria, including having a chain that can be resolved. The bigger problem though is that this is the opposite of how C# (and all object oriented languages) work.

**Reply by JimBobSquarePants** [reply]
> The bigger problem though is that this is the opposite of how C# (and all object oriented languages) work.

That’s FUD.

**Reply by TahirAhmadov** [reply]
Oh it's even worse. For something like:
```cs
SomeType x = new().Prop1.Prop2;
```
`SomeType` is in scope, no problem; it's `Prop1` property is of type `AnotherType`, which is not in scope. This is perfectly legal in C# - you don't need to have `AnotherType` in scope to use it in an expression like this. So, to find `Prop2`, the compiler would need to look for members in all types available in this project, not just those in scope. That would be the most expensive text search ever - we're talking about hundreds of thousands of members.

**Reply by CyrusNajmabadi** [reply]
> Complex or not I strongly believe that the lack of support for this pattern delivers a very poor user experience for developers new to the language.

I don't think it's fair to say 'complex or not'. :)   we have to balance these things out.  If adding a language feature requires enormous complexity/cost, then that's really got to be part of the assessment of things. :)

**Reply by JimBobSquarePants** [reply]
It's still poor UX which should be the most important factor when designing a language feature (_and something that has not been addressed in this thread_).

New developers (_and developers new to C#_) generally don't care how the compiler works or how complex features are to implement. They want consistency and discoverability.

From a natural language perspective 

```c#
T t = new();
T t = new().Chained();
```
are identical statements up to the method expression. That's undeniable. 

Based on that a developer would (_and should_) expect the following rule. 

`new()` MUST be equal to `T`.

Thus any of those chained extensions must be members of `T`.

Don't get me wrong here. I'm a huge fan of target-typed `new` expressions and I understand that language design (_and implementation_) is complex but from a UX perspective this is uncharacteristically unpleasant.

**Reply by 333fred** [reply]
> From a natural language perspective
>
> ```cs
> T t = new();
> T t = new().Chained();
> ```
>
> are identical statements up to the method expression. That's undeniable.

This is not true. Consider the following:

```cs

T t = new(); // Compiles
T t = new().Chained() // Why would this compile? Even if the compiler could figure out you meant `T`, there is no `Chained` for which this works.

class T
{
    public string Chained() => "";
}
```

**Reply by CyrusNajmabadi** [reply]
> It's still poor UX

This is debatable.  :)

You are expressing a strong intuition on your part that this is of course how things should behave.  However, I honestly don't see it the same way.  No part of me sees `<expr>.Something()` and think that the type of `<expr>` is influenced by anything else.  Indeed, i find that super weird as it would somehow mean that both the outer ambient type was complimentary with both the return type of `.Something()` and the *receiver-type* of `.Something()`.  That itself is very rare and not at all something i would think would necessarily be natural, obvious or desirable to encode.

> are identical statements up to the method expression. That's undeniable.

Syre.  But that's just saying that `new()` is equal to `new()`.  But it is *entirely* the surrounding context which makes these different as `new()` *by definition* it target-typed.  So the surrounding `method expression` (which we call a Member-Access + Invocation-Expression) absolutely changes things here :)

>  but from a UX perspective this is uncharacteristically unpleasant.

I genuinely disagree.  I think it's sensible and appropriate.  *If*, though, the BCL were designed entirely fluent methods where the receiver and return type were the same, we might consider codifying this concept.  But it really isn't, so i'm highly skeptical around having the language view things in this fashion.

**Reply by CyrusNajmabadi** [reply]
@333fred and i are on the same page on this.  There is no general correspondence between return-type and receiver-type.  And it's not intuitive to me that the language would somehow assume that there was.  Indeed, it's just not at all clear what hte rules would even be here.  Consider something like:

```c#
T t = new().Chained() + new().Chained();
```

Would this work?  Why or why not?  Would it work through indexing expressions (like `new()[0]`)?    What about `T t = new()++`?

It might be valuable to even express waht you think the langauge rules would be here that would make this work out.

**Reply by JimBobSquarePants** [reply]
I've stated something clearly and undeniably true and yet it's been casually dismissed. I won't lie, it's infuriating.

> From a natural language perspective
> 
> ```c#
> T t = new();
> T t = new().Chained();
> ```
> are identical statements up to the method expression. That's undeniable.

[Natural Language](https://dictionary.cambridge.org/dictionary/english/natural-language)

>language that has developed in the usual way as a method of communicating between people, rather than language that has been created, for example for computers:

For left to right languages that means that syntax is parsed from left to right. We don't read ahead for understanding we work in one direction and understanding is built incrementally.

>This is debatable. :)

I'm telling you, as a customer, that the current limitation is confusing and poor UX. That shouldn't be treated as debatable. I'm telling you categorically that, as a result of this behavioral difference, I (_and others_) are **having a poor user experience**.

>it's just not at all clear what hte rules would even be here

The expected rules are obvious (and stated by the OP) and I really do not understand why there is confusion and why individuals in this thread insist on branching out with unrelated examples. 

>It appears the compiler is attempting to parse the entirety of the expression and infer the type from that, whereas to my mind it should be parsing and resolving DefaultHttpContext context = new() first, then evaluating the chained method call.

They could not have expressed this more clearly in a manner that correlates with my description of natural language yet were met with 

>Why would it do that? 

Which reads, unpleasantly, as distaining. 

Here's the rule. 

**Do EXACTLY the SAME thing for both lines until they change.**

```c#
<-- SAME  -->
|           | 
|T t = new()|;
|           |

|           |
|T t = new()|.Chained();
|           |
```

**Reply by CyrusNajmabadi** [reply]
> I've stated something clearly and undeniably true and yet it's been casually dismissed.

No one has been casual.  We've just given our beliefs about what is going on here.

**Reply by TahirAhmadov** [reply]
> For left to right languages that means that syntax is parsed from left to right. We don't read ahead for understanding we work in one direction and understanding is built incrementally.

Exactly! And your proposal would make the compiler interpret that line of code **right to left**. That's the problem.

> I've stated something clearly and undeniably true and yet it's been casually dismissed. 

You completely ignored the points I raised which clearly demonstrate how your proposal just won't work. **Not** "oh it's difficult and we can't spend that much time on it", but "it's mathematically/logically impossible to accomplish".

> I won't lie, it's infuriating.

Here, I differ. I'm not infuriated one bit.

**Reply by CyrusNajmabadi** [reply]
> Here's the rule.
> 
> Do EXACTLY the SAME thing for both lines until they change.

*Here's the thing though*

Our language does not work this way. There are many cases *since C# 1.0* where that above intuition you have simply does not hold.

As a simple example:

```
Foo(x, y, ...   // vs
Foo(x, y, z, ...
```

Here, 'x' and 'y' do not do the exact same thing or have the same meaning as they can trivially be affected by what *follows after* :)

**Reply by CyrusNajmabadi** [reply]
> For left to right languages that means that syntax is parsed from left to right. We don't read ahead for understanding we work in one direction and understanding is built incrementally.

FWIW this is often not true with programming languages.  It is also often not true with natural languages either.  

> I'm telling you, as a customer, that the current limitation is confusing and poor UX. That shouldn't be treated as debatable.

It can and should be debated.  And we will not shy away from debating things here.  One person's report is not gospel here.  Furthermore, by this definition, all things have poor UX, as we can always find a customer that finds it confusing.

>  I'm telling you categorically that, as a result of this behavioral difference, I (and others) are having a poor user experience.

I have not disagreed with that.  What i said is that this is debatable about if this is overall poor UX.  

For example, if you have a UX that 999,999 people find fine, and 1 person does not.  It is certainly true that that 1 person is having a poor experience.  But no lang designer would consider that feature to be poor UX.

> The expected rules are obvious (and stated by the OP) 

The rule is non-obvious to me.  And I asked as we would need a spec on this and it's important to have consensus about what that spec would state.  Please keep in mind the .net code of conduct here:

> Demonstrating empathy and kindness toward other people
> Being respectful of differing opinions, viewpoints, and experiences
> Giving and gracefully accepting constructive feedback

I have asked for something that would help me out.  Please accept that I need that.

> and I really do not understand why there is confusion and why individuals in this thread insist on branching out with unrelated examples.

Because they're not unrelated *to us*.  They are part of the greater domain of the language we would need to explore and understand to intuit, spec and implement this feature :)

> Which reads, unpleasantly, as distaining.

It isn't.  It's a totally reasonable and respectful question.  This forum is explicitly intended for language ideas and debate.  We stated that *explicitly* in our readme.  Having someone ask a basic question like this is not only acceptable, but totally reasonable and normal here.  

You should consider that not everyone will feel the same way as everyone else.  And the purpose of this interaction is to allow others to better understand you and for you to better understand them.

Thanks! :)

**Reply by CyrusNajmabadi** [reply]
>  I won't lie, it's infuriating.

Please be respectful and understand that not everyone feels the same as you.  This sort of conversation is an age old debate around ways of interpretting syntax.  For example, you still see those FB quizzes going: 

```
// What does the following equal?!
x = 3 + 4 * 5;
```

Some people follow PEMDAS and say `23` (e.g. `3 + (4 * 5)`).  Others take a left-to-right reading and say `35` (e.g. `(3 + 4) * 5`).  

Getting infuriated because people have a different interpretation than you is not helpful.  Instead, please work to understand *why* there is a difference in opinion, and if there's a common ground or way to propose things that might meet the desires and views of both groups.

--

Pulling this back to C#, we have never been a "the meaning stays the same as you add tokens to the right" language.   Like *most* math/programming/logic languages, we have concepts of operators, precedence, associativity, parse trees, and the like.  In these worlds it is *more often* the case that thsi rule does *not* apply and that tokens/nodes to the right change the meaning of what is to the left. 

So, while you are totally within your right to wish the language behaved in the manner you've expressed, others are well within their right to point out that C# has never worked this way and that reconciling these differences will be challenging here.

**Reply by theunrepentantgeek** [reply]
> For left to right languages that means that syntax is parsed from left to right. We don't read ahead for understanding we work in one direction and understanding is built incrementally.

I disagree with your initial premise - reading ahead and rereading preceeding text is extremely common.

Even if I accepted your premise, clauses that come later can dramatically revise the meaning of what came before.

If you'll forgive a couple of slightly contrived examples ...

Guests may not speak
_vs_
Guests may not speak until spoken to by the Queen.

In some times and places, failing to speak when spoken to by the sovereign could quite literally be a fatal mistake.

Ignite the fuse
_vs_
Ignite the fuse only after ensuring the blast site is clear.

**Reply by CyrusNajmabadi** [reply]
Note.  As to your point about "Natural Language".  I'll point out that the very definition you linked to mentioned exactly:

> language that has developed in the usual way as a method of communicating between people, **rather than language that has been created, for example for computers**:

Important part in bold.  Computer languages are not natural languages.  Nor is it necessarily appropriate or desirable to try to model them that way.  Indeed, 'natural languages' suffer enormous from problems that we try to keep very far away from programming languages :)  

I can state pretty firmly that:

1. it is a non-goal for C# to either be, or ape, "natural languages".
2. "natural languages" *themselves* have "terrible UX" :)  

'2' is not a joke.  Think about how much confusion and ambiguity arises in natural languages.  It's a genuine significant problem, which is why we tend to keep that world and the world of programming languages very separated.

**Reply by JimBobSquarePants** [reply]
I'm tapping out here. There's no point engaging with individuals so well practiced in condescension.

**Reply by CyrusNajmabadi** [reply]
@JimBobSquarePants your behavior is not acceptable. You were not shown condescension.  People raised appropriate points against the proposal here and the points you yourself made. 

If you cannot treat the members of this community with appropriate respect and tolerance for differing opinions then you will not be welcome here in the future.

**Reply by DavidArno** [reply]
@JimBobSquarePants,

> I'm telling you, as a customer, that the current limitation is confusing and poor UX. That shouldn't be treated as debatable.

The challenge here for the language team is that - also as a customer - I'm telling them that as it is feels right and I'd find your suggested approach deeply confusing. So they have customers telling them conflicting things. Your suggestion is logical, I just think the current approach is more logical. Your implied "it's confusing for all (new) developers" most certainly is debatable. I think you are wrong. And you are of course entitled to think me wrong too.

**Reply by jmarolf** [reply]
*Attempting to summarize the problem in my own words (as I understand it)*

Target-typed new does not play well with fluent extensions. 

as the original poster showed you can do this:

```csharp
DefaultHttpContext context = new DefaultHttpContext(); // ugh so much repetition
var context = new DefaultHttpContext(); // better but what if you hate var or are not in a variable declaration context?
DefaultHttpContext context = new(); // nice
```

Of those three forms the first two play well with fluent extension / fluent apis

```csharp
DefaultHttpContext context = new DefaultHttpContext().Some().Cool().Configuration().Build();
var context = new DefaultHttpContext().Some().Cool().Configuration().Build();
```

but target typed new does not.
```csharp
DefaultHttpContext context = new().Some().Cool().Configuration().Build(); // compile error
```

Supposing target typed new did work in this location there would be a few benefits (imho).

1. [Minimal apis](https://docs.microsoft.com/aspnet/core/fundamentals/minimal-apis) are the way we would like ASP.NET apps to be (might me reaching here but we certainly expect minimal apis to be the way people new to .NET get started) and in general our API design is trending towards more fluent apis not less.  So, making more patterns work in fluent apis means newcomers are less likely to write something, modify something, or copy/paste something and get a random compiler error that they will need to inspect.
2. While `var` was met with resistance and there is still a 50/50 split in the community on its usage, target-typed-new was met with fairly universal praise by comparison. For the half of .NET developers that do not use `var` (or cannot) this is a welcome feature and having different contexts in which it can work compared to `var` feels like a take-back or a bit of unfairness and favoritism toward `var`. I am certain I am attributing more intense feelings about this than what most people actually have but I still think that if target typed new could work more places it would be a net positive for folks who do not use `var` in their codebase.

Probably other positives I am missing but now we need to move on to the sadder part of this conversation: what would be the language consequences of doing this.

So let's walk down the path of what this could look like. First let's just skip over the hard part and assume that this complies now:

```csharp
DefaultHttpContext context = new().Some().Cool().Configuration().Build(); // all good;
```

In this case each extension method (or member method) returns the same type `DefaultHttpContext` and Chuck or Cyrus or some other genius sat down and made it all work. 

So, if we go back to the original code and something like this

```csharp
DefaultHttpContext context = new().Some().Cool().Configuration().Build().AsNotDefaultHttpContext();
```

We would want to error, and let's just skip over how we do that and assume we do a great job and produce something like this: 

```csharp
DefaultHttpContext context = new()
    .Some()
    .Cool()
    .Configuration()
    .Build()
    .AsNotDefaultHttpContext(); //Cannot convert from 'NotDefaultHttpContext' to 'DefaultHttpContext'
```

but then what if we did this:
```csharp
DefaultHttpContext context = new()
    .Some()
    .Cool()
    .Configuration()
    .Build()
    .AsNotDefaultHttpContext()
    .JustKiddingBackToDefaultHttpContext(); // no error?
```

To get to the point: Do we want to say that `new()` is whatever the left-hand assignment is or the final expression result? Consider this:

```csharp
A? context1 = new A().Configure().Make().Build(); // Cannot implicitly convert type 'C' to 'A'
A? context2 = new B().Configure().Make().Build(); // No Error

public class A { }

public class B { }

public class C { }

public static class ExensionMethods
{
    public static B? Configure(this A? a) => null;
    public static C? Make(this B? a) => null;
    public static C? Build(this C? a) => null;

    public static A? Configure(this B? a) => null;
    public static B? Make(this A? a) => null;
    public static A? Build(this B? a) => null;
}
```

Based on this example I assume we want the same error to appear in this case:

```csharp
A? context1 = new ().Configure().Make().Build(); // Cannot implicitly convert type 'C' to 'A'
```

versus something like this:
```csharp
A? context1 = new ().Configure().Make().Build(); // no error because we are constructing B
```

So the general rule (that would need to be formalized) is: 
- Target typed new is allowed in assignments where the resulting type of the right-hand expression exactly matches the type in the variable declarator _when we substitute the left hand types constructor into the target-typed-new expression_.

```csharp
A? context = new().Something(); // Error no matching constructor
A? context = new(Id: 42).SomethingElse(); // Error expression doesn't result in type convertible to 'A'
A? context = new(Id: 42).Something(); // Fine
record A(int Id);
record B;
public static class ExensionMethods
{
    public static A? Something(this A? a) => null;
    public static B? SomethingElse(this A? a) => null;
}
```

I _think_ I am ok with this. This isn't how target-type new works at all today under the hood. But hand-waving all that away I think carving out a special case for fluent methods would be fine. Separate question would be where this falls on the priority list, how this should actually be spec'd out and where there is a case I am not thinking of where the `new()` would be ambiguous because of the way the right hand expression is constructed (though my answer there is to just do what we do in ternary expressions and fail to compile).

**Reply by TahirAhmadov** [reply]
@jmarolf so that would be a very strange set of limitations; `new()` is allowed as a first step after `A context = `, and it means `new A()`; then, any number of members can be called on the right, but the end result also has to be `A`? That introduces a programming style not in use in any other language that I know of. Furthermore, while I guess that _could_ work, but the question remains - why? How is that better than `var context = new A().Configure().Make().Build();`?

**Reply by tannergooding** [reply]
@TahirAhmadov, for `var context = new A().Configure().Make().Build();` the types are all clear because `new A()` can only be `A` and therefore each subsequent method call can only resolve to "one thing".

However for `var context = new().Configure().Make().Build();`, there is no way to statically determine what `new()` returns. You could technically iterate through all members that expose an empty construct and a `Configure()` method, but that's also trivially subject to breaking changes and conflicts. It would also be "expensive" to resolve as was pointed out above.

You run into similar problems with `SomeType context = new().SomeMethod()`. You could assume `SomeType` is "correct" here, but that is also potentially problematic because there is nothing definitively stating that `SomeMethod()` is a proper "fluent" API.

The language is left "guessing" at what you meant and that's generally bad and leads to bugs.

**Reply by TahirAhmadov** [reply]
> However for `var context = new().Configure().Make().Build();`, there is no way to statically determine what `new()` returns. 

And why would anybody expect that to work? What's stopping the coder to write `var context = new A().Configure().Make().Build();` ??

**Reply by jmarolf** [reply]
> so that would be a very strange set of limitations; new() is allowed as a first step after A context = , and it means new A(); then, any number of members can be called on the right, but the end result also has to be A?

Yes sorry, my intent was to not post non-compiling code. You can totally do `A? context = new ()` and it works without issues but I agree that most people do not (and should not) do that unless `context` is intended to be a mutable value that you may assign null to. In reality a developer would probably do `A context =  new().<Some Fluent Expression>` and then if the result of that was `A?` they would get a nullable warning, but I didn't want to write that explanation out, sorry for the oddness.

**Reply by tannergooding** [reply]
> And why would anybody expect that to work?

That's my point. There isn't a huge difference between: `var context = new().Configure().Make().Build();` and `T context = new().Configure().Make().Build();` both require the compiler to do additional and effectively "magic" resolution.

Given that expression, there is no way to determine what `new()` is meant to create because there is no target type for it. If you break the above expression apart you are effectively doing:
```csharp
var tmp1 = new();
var tmp2 = tmp1.Configure();
var tmp3 = tmp2.Make();
T context = tmp3.Build();
```

Where `tmp1`, `tmp2`, and `tmp3` all exist as "intermediates" on the execution stack in IL and so its invalid because the first step of the expression is `new()` where the target type is not given.

Changing this to `new A()` allows it to work because now the type is "well-defined", as would any other combination of expressions where the first type can be determined.

**Reply by jmarolf** [reply]
> However for var context = new().Configure().Make().Build();, there is no way to statically determine what new() returns.

My understanding is that `new()` is implemented as a new expression that can be convertable to the left-hand type by calling the right constructor (though I need to actually go look at the implementation to check how correct this is). So, like what we do in other cases in the compiler, we could change the rules so (after all our existing lookup rules fail) 
1. we look at the type on the left hand side
2. attempt to directly substitute the types constructor replacing the new expression (which is how I think the original poster thought this feature worked) 
3. If the lookup rules allow us to find a suitable constructor, attempt to bind the entire right-hand side expression
4. If the resultant type of the expression is convertible to the left hand side allow that constructor substitution.

One problem with the approach (aside from the implementation complexity or having two different mechanism to resolve new() expressions)  is that I think error recovery in lambdas might be made far worse with this rule. Suppose I am typing inside this lambda expression. 

```csharp
`List<T> context = new().Select(x => )`
```

We need to substitute and bind things to even know what `x` is but we can't resolve whether `new()` is valid until we have returned from the lambda. We could do the work here to try and make error recovery as good as possible but now things aren't simple at all if we want to developer experience to be good.

Considering that the point of this (as I understand it) is to make fluent apis better I don't know this would result in a good experience if those fluent apis contained lambdas.

**Reply by CyrusNajmabadi** [reply]
i believe the customer request though is to rewrite the above as:

```c#
T tmp1 = new();
var tmp2 = tmp1.Configure();
var tmp3 = tmp2.Make();
T context = tmp3.Build();
```

Note: i do not like this *for many reasons*.  But it a *potentially* viable path.

Note: one way to even accomplish this *syntactically* would be to make a break and have the tree here become:

```
           implicit-obj-creation-and-access (this is a new node/concept)
          / \
     new().  RHS
            / \
           .   Build()
          / \
Configure()  Make()
```

Thus `new()` would be bound first in the target typed context of it's location.

This would be similar to how we parse `?.` making it so that the LHS evals first, is checked against null or not, and then the entire RHS is evaluated against that.


---

Again, i do not like it *at all* :)   But it's a formalization that could make this work, albeit with enormous contortions and implications for compat.

**Reply by TahirAhmadov** [reply]
> > so that would be a very strange set of limitations; new() is allowed as a first step after A context = , and it means new A(); then, any number of members can be called on the right, but the end result also has to be A?
> 
> Yes sorry, my intent was to not post non-compiling code. You can totally do `A? context = new ()` and it works without issues but I agree that most people do not (and should not) do that unless `context` is intended to be a mutable value that you may assign null to. In reality a developer would probably do `A context = new().<Some Fluent Expression>` and then if the result of that was `A?` they would get a nullable warning, but I didn't want to write that explanation out, sorry for the oddness.

Just a misunderstanding, I didn't mean `A?` (nullable), I was asking a question. Nullability is irrelevant here (at least, for now).

My question is, given these strict limits, how is this syntax:
```cs
Class c = new().Configure().Build().Make(); // new() is made to be new Class(), and Make() returns Class
```
Any better than this?
```cs
var c = new Class().Configure().Build().Make(); // Make() returns Class
```
And obviously we all understand that the below is not just impossible but TOTALLY impossible:
```cs
var c = new().Configure().Build().Make();
```

**Reply by TahirAhmadov** [reply]
> > And why would anybody expect that to work?
> 
> That's my point. ......

Please see my post above; I'm pretty sure we are in agreement that OP's proposal is at worst impossible and at best not useful.

**Reply by tannergooding** [reply]
> i believe the customer request though is to rewrite the above as:

Right. I was just trying to explain why the code doesn't make sense/work today.

That being said, I also don't like this and don't think this is intuitive, particularly given how expressions work in C# today. But that's my personal opinion here.

**Reply by CyrusNajmabadi** [reply]
> how is this syntax: ... Any better than this?

One could be used in a field, the other could not.

**Reply by TahirAhmadov** [reply]
> One could be used in a field, the other could not.

I knew somebody would bring it up! Well, first of all, code like this doesn't belong in a class initializer - it may legally, but it's highly unlikely to actually happen. Second, even in a field, as much as we all hate typing, `Class _c = new Class().Configure().Build().Make();` is all of N-3 more letters to type, where N is the name of the class. Given that a) it's a rare form and b) it only saves a few keystrokes (or none at all with autocomplete), I think the cost is huge and benefit is almost non-existent.

**Reply by CyrusNajmabadi** [reply]
> Well, first of all, code like this doesn't belong in a class initializer - it may legally, but it's highly unlikely to actually happen

I don't really have any confidence in that statement.

**Reply by TahirAhmadov** [reply]
> I don't really have any confidence in that statement.

I do. :)

**Comment by DavidArno** [comment]
> I have checked https://docs.microsoft.com/dotnet/csharp/language-reference/proposals/csharp-9.0/target-typed-new but can't see anything that would imply this shouldn't work.

In the miscellaneous section, it states it will not work:

"*It is disallowed when ... in a member access (`new().field`) ...*"
